Features/ManageabilityPkg/Library/ManageabilityTransportSerialLib/Common/SerialCommon.c

```diff
@@ -0,0 +1,607 @@
+
+  @param[in]      Character           Character data.
+
+  @retval         Character           Return escaped character if it is.
+                                      Otherwise, return the given one.
+**/
+static
+UINT8
+IpmiSerialGetEscapedCharacter (
+  UINT8  Character
+  )
+{
+  UINT8  Index;
+
+  for (Index = 0; Index < mNumberOfEscapeCharacters; Index++) {
+    if (mIpmiSerialSpecialChar[Index].Character == Character) {
+      return mIpmiSerialSpecialChar[Index].Escape;
+    }
+  }
+
+  return Character;
+}
+
+/**
+  Return unescaped character for the given one
+
+  @param[in]      Character           Character data.
+
+  @retval         Character           Return unescaped character if it is.
+                                      Otherwise, return the given one.
+**/
+static
+UINT8
+IpmiSerialGetUnescapedCharacter (
+  UINT8  Character
+  )
+{
+  UINT8  Index;
+
+  for (Index = 0; Index < mNumberOfEscapeCharacters; Index++) {
+    if (mIpmiSerialSpecialChar[Index].Escape == Character) {
+      return mIpmiSerialSpecialChar[Index].Character;
+    }
+  }
+
+  return Character;
+}
+
+/**
+  Parse IPMI Serial incoming data in basic mode format to IPMI message
+
+  @param[in]        Buffer                Data buffer with data byte escaped.
+  @param[in]        BufferSize            Data buffer size.
+  @param[out]       ResponseData          Data buffer without data byte escaped.
+  @param[in, out]   ResponseDataSize      Data buffer size.
+
+  @retval           EFI_SUCCESS           The command byte stream was successfully
+                                          submit to the device and a response was
+                                          successfully received.
+  @retval           EFI_PROTOCOL_ERROR    The command byte stream cannot encode
+                                          successfully.
+  @retval           EFI_OUT_OF_RESOURCES  The resource allocation is out of resource or
+                                          data size error.
+**/
+EFI_STATUS
+IpmiSerialParseIncomingBuffer (
+  IN        UINT8   *Buffer,
+  IN        UINT32  BufferSize,
+  OUT       UINT8   *ResponseData,
+  IN OUT    UINT32  *ResponseDataSize
+  )
+{
+  UINT8         *ResponsePtr;
+  UINT8         Character;
+  UINT8         UnescapedCharacter;
+  UINT32        Count;
+  UINT32        Index;
+  STATIC UINT8  CtxState  = MSG_IDLE;
+  STATIC UINT8  CtxEscape = 0;
+
+  if ((BufferSize == 0) || (Buffer == NULL)) {
+    DEBUG ((DEBUG_ERROR, "Buffer Data insufficient. \n"));
+    return EFI_OUT_OF_RESOURCES;
+  }
+
+  if ((ResponseDataSize == NULL) ||
+      (BufferSize > *ResponseDataSize) ||
+      (ResponseData == NULL))
+  {
+    DEBUG ((DEBUG_ERROR, "Response Data insufficient. \n"));
+    return EFI_OUT_OF_RESOURCES;
+  }
+
+  ResponsePtr = ResponseData;
+  Count       = 0;
+  for (Index = 0; Index < BufferSize; Index++) {
+    Character = Buffer[Index];
+    if (Character == BASIC_MODE_START) {
+      // START

```

Comment:
A check is required before assigning to CtxState.
it should be idle state.

Reply:
Acked.